languages/csharp/exercises/concept/resource-cleanup/ResourceCleanupTests.cs

```diff
@@ -0,0 +1,111 @@
+using System;
+using System.Reflection;
+using Xunit;
+using template;
+
+public class ResourceCleanupTests
+{
+    [Fact]
+    public void Write_good()
+    {
+        var db = new Database();
+        var orm = new Orm(db);
+        orm.Write("good write");
+        object[] actual = {db.DbState, db.lastData};
+        Assert.Equal(new object[] { Database.State.DataWritten, "good write"}, actual);
+    }
+
+    [Fact /*(Skip = "Remove this Skip property to run this test")*/]
+    public void Write_bad()
+    {
+        var db = new Database();
+        var orm = new Orm(db);
+        orm.Write("bad write");
+        object[] actual = {db.DbState, db.lastData};
+        Assert.Equal(new object[] { Database.State.Closed, "bad write"}, actual);
+    }
+
+    [Fact /*(Skip = "Remove this Skip property to run this test")*/]
+    public void Commit_good()
+    {
+        var db = new Database();
+        var orm = new Orm(db);
+        orm.Write("good commit");
+        orm.Commit();
+        object[] actual = {db.DbState, db.lastData};
+        Assert.Equal(new object[] { Database.State.Closed, "good commit"}, actual);
+    }
+
+    [Fact /*(Skip = "Remove this Skip property to run this test")*/]
+    public void Commit_bad()
+    {
+        var db = new Database();
+        var orm = new Orm(db);
+        orm.Write("bad commit");
+        orm.Commit();
+        object[] actual = {db.DbState, db.lastData};
+        Assert.Equal(new object[] { Database.State.Closed, "bad commit"}, actual);
+    }
+
+    [Fact /*(Skip = "Remove this Skip property to run this test")*/]
+    public void Disposable()
+    {
+        var db = new Database();
+        var orm = new Orm(db);
+        orm.Write("good data");
+        bool disposable = orm is IDisposable;
+        if (disposable)
+        {
+            typeof(Orm).InvokeMember("Dispose",
+                BindingFlags.Public | BindingFlags.Instance | BindingFlags.InvokeMethod,
+                null, orm, null);
+        }
```

Comment:
I assume you've added this to deal with the case of students removing the `IDisposable` interface and the `Dispose` method? Personally, I think it is perfectly fine to just have it be a compile warning, as we already provide a compiling stub.

```suggestion
        orm.Dispose()
```

We could have a separate test method to verify that `IDisposable` is implemented that does:

```csharp
Assert.IsAssignableFrom<IDisposable>(typeof(Orm));
```

Reply:
>  assume you've added this to deal with the case of students removing the IDisposable interface

No.  My intention was that the student would have to implement `IDisposable` on the `Orm` class.  And as we've discussed before I am not keen to have compile errors in the code we send out.

I think there is something to be said for making the student see that if you are using an object that implements `IDisposable` then there's a fair chance that you will have to implement it on your class.

Reply:
But the stub for the `Orm` class already implements `IDisposable` (with a `NotImplementedException` of course). Is that incorrect then?

Reply:
Copy+Paste error, yuk.  Task 4, _Ensure that the database is cleaned up correctly_ is where `IDisposable` should be added.

Reply:
Understood.